java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -127,15 +127,7 @@ class HttpString extends Expr {
   HttpString() {
     this instanceof HttpStringLiteral
     or
-    this.(VarAccess).getVariable().getAnAssignedValue() instanceof HttpStringLiteral
-    or
     concatHttpString(this.(AddExpr).getLeftOperand(), this.(AddExpr).getRightOperand())
-    or
-    concatHttpString(this.(AddExpr).getLeftOperand().(AddExpr).getLeftOperand(),
-      this.(AddExpr).getLeftOperand().(AddExpr).getRightOperand())
-    or
-    concatHttpString(this.(AddExpr).getLeftOperand(),
-      this.(AddExpr).getRightOperand().(AddExpr).getLeftOperand()) // First two elements of a string concatenated from an arbitrary number of elements.
   }
```

Comment:
This could cause false positives for something like `"http://" + ("localhost:" + port)` though I assume something like that is rather unlikely so considering it might not be worth it.

Reply:
As the implementation doesn't validate the port, any character after the host string is fine. I've confirmed the query works with `String host="localhost:"; new URL("http://"+host...)` and `new URL("http://", host, ...)`.

Please let me know if I miss something. I won't close this conversation until your confirmation.

Reply:
What I mentioned here is related to @aibaars' suggestion described in https://github.com/github/codeql/pull/3976#discussion_r463473459.

My comment was about the way the string is concatenated. The expression `"http://" + ("localhost:" + port)` is an `AddExpr` with:
- left: "http://"
- right: AddExpr

So even though it is clear here that the host is `localhost`, `concatHttpString` would hold since `getRightOperand()` returns an `AddExpr` and not a compile time constant.

It is unlikely that such a parenthesized string concatenation is used, but to cover it, something like the following could be used:
```ql
Expr getLeftmostConcatOperand(Expr expr) {
  if expr instanceof AddExpr then (
    result = getLeftmostConcatOperand(expr.(AddExpr).getLeftOperand())
  ) else result = expr
}

...

class HttpString extends Expr {
  HttpString() {
    this instanceof HttpStringLiteral
    or
    concatHttpString(this.(AddExpr).getLeftOperand(), getLeftmostConcatOperand(this.(AddExpr).getRightOperand()))
  }
}
```
This would also cover contrived cases such as `"http://" + ((("localhost:" + port) + query) + fragment)`.

Reply:
Thanks @Marcono1234 for the detailed instructions and I've updated the query. Now the test passes with the following:
```
String host = "LOCALHOST:";
		HttpURLConnection conn = (HttpURLConnection) new URL("http://"+(((host+"/rest/getuser.do")+"?uid=abcdx"))).openConnection();
```